### PR TITLE
Improve subs

### PR DESCRIPTION
**Pull request by aivit**
**Changes**
Improve subtitles by changing default dont to universal system font stack(the same one github uses) and improving stroke for uniform option.

**Before/after video**
``00:00-00:17 = before(current jellyfin)``
``00:18-00:32 = after(pull request version)``
https://youtu.be/XpWnDU5S_6k

**Screenshot**
Screen Shot 2022-04-29 at 8 36 47 AM

### CODE REVIEW

**Comment by cvium** [comment]
When making visual changes it's very helpful if you include before and after pics.

**Comment by dannymichel** [comment]
> When making visual changes it's very helpful if you include before and after pics.

Good point
Screen Shot 2022-04-29 at 8 32 46 AM
Screen Shot 2022-04-29 at 8 33 40 AM

**Comment by dannymichel** [comment]
**Before/after video**
``00:00-00:17 = before(current jellyfin)``
``00:18-00:32 = after(pull request version)``
https://youtu.be/XpWnDU5S_6k

**Comment by thornbill** [comment]
I think the font changes are fine, but the shadow looks a bit harsh. I believe @dmitrylyzo had a suggestion for a softer shadow.

**Comment by dannymichel** [comment]
> I think the font changes are fine, but the shadow looks a bit harsh. I believe @dmitrylyzo had a suggestion for a softer shadow.

Thanks
I mean sure I guess I can do that
I just took the css directly from Netflix to be perfectly honest

**Comment by dmitrylyzo** [comment]
The problem is that `px` is always `px`. With a small font, the shadow starts to look oversized.

As has been said, to show the UI changes, it is better to add "before/after" pics. The video is good for demonstrating changes in behavior (or issues).

**Before**
_These are Smaller, Normal and Extra Large. But they are scaled down in the preview window, so you can skip the first line, and the rest are equal to Smaller and Normal._
subs-old
**After (px - yours)**
subs1

I tried to use `em` as `1px -> 0.02em` and added more layers (which I don't like because it is more things to draw). In my test, I also removed `0,0`.

**After (em)**
subs-em

Your pattern:
```
 22222 
3211123
3210123
3211123
 22222 
```
My pattern:
```
 33333 
3222223
3211123
321 123
3211123
3222223
 33333 
```

You could try (with `1px -> 0.03em`):
```
 222 
21112
21 12
21112
 222 
```
_Actually, I just tried it and it looks acceptable._

Moreover, I think we can use `#000` for short.

**Comment by dannymichel** [comment]
I'll change it to em

**Comment by dannymichel** [comment]
> The problem is that `px` is always `px`. With a small font, the shadow starts to look oversized.
> 
> As has been said, to show the UI changes, it is better to add "before/after" pics. The video is good for demonstrating changes in behavior (or issues).
> 
> **Before** _These are Smaller, Normal and Extra Large. But they are scaled down in the preview window, so you can skip the first line, and the rest are equal to Smaller and Normal._ subs-old **After (px - yours)** subs1
> 
> I tried to use `em` as `1px -> 0.02em` and added more layers (which I don't like because it is more things to draw). In my test, I also removed `0,0`.
> 
> **After (em)** subs-em
> 
> Your pattern:
> 
> ```
>  22222 
> 3211123
> 3210123
> 3211123
>  22222 
> ```
> 
> My pattern:
> 
> ```
>  33333 
> 3222223
> 3211123
> 321 123
> 3211123
> 3222223
>  33333 
> ```
> 
> You could try (with `1px -> 0.03em`):
> 
> ```
>  222 
> 21112
> 21 12
> 21112
>  222 
> ```
> 
> _Actually, I just tried it and it looks acceptable._
> 
> Moreover, I think we can use `#000` for short.

I changed it to em and it looks much better small or large

File: src/components/subtitlesettings/subtitleappearancehelper.js [hunk 1]
```diff
@@ -30,13 +30,13 @@ function getTextStyles(settings, preview) {
 
     switch (settings.dropShadow || '') {
         case 'raised':
-            list.push({ name: 'text-shadow', value: '-1px -1px white, 0px -1px white, -1px 0px white, 1px 1px black, 0px 1px black, 1px 0px black' });
+            list.push({ name: 'text-shadow', value: '-0.04em -0.04em  #FFF, 0px -0.04em #FFF, -0.04em 0px #FFF, 0.04em 0.04em #000, 0px 0.04em #000, 0.04em 0px #000' });
```
Comment (dmitrylyzo) [inline src/components/subtitlesettings/subtitleappearancehelper.js:33]: Mostly the lower case is used. Also redundant space.
```suggestion
            list.push({ name: 'text-shadow', value: '-0.04em -0.04em #fff, 0px -0.04em #fff, -0.04em 0px #fff, 0.04em 0.04em #000, 0px 0.04em #000, 0.04em 0px #000' });
```
I'm not sure about this change (I don't like `Raised`/`Depressed`), but at least logically it looks right.

File: src/components/subtitlesettings/subtitleappearancehelper.js [hunk 2]
```diff
@@ -30,13 +30,13 @@ function getTextStyles(settings, preview) {
...
             break;
         case 'depressed':
-            list.push({ name: 'text-shadow', value: '1px 1px white, 0px 1px white, 1px 0px white, -1px -1px black, 0px -1px black, -1px 0px black' });
+            list.push({ name: 'text-shadow', value: '0.04em 0.04em #FFF, 0px 0.04em #FFF, 0.04em 0px #FFF, -0.04em -0.04em #000, 0px -0.04em #000, -0.04em 0px #000' });
```
Comment (dmitrylyzo) [inline src/components/subtitlesettings/subtitleappearancehelper.js:36]: Mostly the lower case is used.
```suggestion
            list.push({ name: 'text-shadow', value: '0.04em 0.04em #fff, 0px 0.04em #fff, 0.04em 0px #fff, -0.04em -0.04em #000, 0px -0.04em #000, -0.04em 0px #000' });
```
I'm not sure about this change (I don't like `Raised`/`Depressed`), but at least logically it looks right.

**Comment by dannymichel** [comment]
Did the cleanup, removed the bold

File: src/components/subtitlesettings/subtitleappearancehelper.js [hunk 3]
```diff
@@ -30,13 +30,13 @@ function getTextStyles(settings, preview) {
 
     switch (settings.dropShadow || '') {
         case 'raised':
-            list.push({ name: 'text-shadow', value: '-1px -1px white, 0px -1px white, -1px 0px white, 1px 1px black, 0px 1px black, 1px 0px black' });
+            list.push({ name: 'text-shadow', value: '-0.04em -0.04em #fff, 0px -0.04em #fff, -0.04em #fff, 0.04em 0.04em #000, 0px 0.04em #000, 0.04em #000' });
```
Comment (dmitrylyzo) [inline src/components/subtitlesettings/subtitleappearancehelper.js:33]:
```suggestion
            list.push({ name: 'text-shadow', value: '-0.04em -0.04em #fff, 0px -0.04em #fff, -0.04em #fff, 0.04em 0.04em #000, 0px 0.04em #000, 0.04em 0px #000' });
```

File: src/components/subtitlesettings/subtitleappearancehelper.js [hunk 4]
```diff
@@ -30,13 +30,13 @@ function getTextStyles(settings, preview) {
...
             break;
         case 'depressed':
-            list.push({ name: 'text-shadow', value: '1px 1px white, 0px 1px white, 1px 0px white, -1px -1px black, 0px -1px black, -1px 0px black' });
+            list.push({ name: 'text-shadow', value: '0.04em 0.04em #fff, 0px 0.04em #fff, 0.04em #fff, -0.04em -0.04em #000, 0px -0.04em #000, -0.04em #000' });
```
Comment (dmitrylyzo) [inline src/components/subtitlesettings/subtitleappearancehelper.js:36]:
```suggestion
            list.push({ name: 'text-shadow', value: '0.04em 0.04em #fff, 0px 0.04em #fff, 0.04em #fff, -0.04em -0.04em #000, 0px -0.04em #000, -0.04em 0px #000' });
```

**Review by dmitrylyzo — COMMENTED** [review]
> Did the cleanup
You overdid it a bit.

The question about the pattern still stands.

**Comment by dannymichel** [comment]
> > Did the cleanup
> 
> The question about the pattern still stands.
The pattern I believe is a preference or standard thing. I see this pattern as a standard on other platforms.

**Comment by dannymichel** [comment]
> > Did the cleanup
> 
> You overdid it a bit.
> 
Changed that to 03

File: src/components/subtitlesettings/subtitleappearancehelper.js (hunk 3 above)
Comment (dannymichel) [reply src/components/subtitlesettings/subtitleappearancehelper.js]: https://github.com/jellyfin/jellyfin-web/pull/3607/commits/7da80ab031249a83e4477c9139429be1ffa24275

File: src/components/subtitlesettings/subtitleappearancehelper.js (hunk 4 above)
Comment (dannymichel) [reply src/components/subtitlesettings/subtitleappearancehelper.js]: https://github.com/jellyfin/jellyfin-web/pull/3607/commits/7da80ab031249a83e4477c9139429be1ffa24275

File: src/components/subtitlesettings/subtitleappearancehelper.js [hunk 5]
```diff
@@ -28,15 +28,27 @@ function getTextStyles(settings, preview) {
...
+            list.push({ name: 'font-weight', value: 'normal' });
+            break;
+        case 'bold':
+            list.push({ name: 'font-weight', value: '600' });
+            break;
+        default:
+            list.push({ name: 'font-weight', value: 'normal' });
+            break;
+    }
+
```
Comment (dmitrylyzo) [inline src/components/subtitlesettings/subtitleappearancehelper.js:42]: Good start, but save it for later - the release branch is impossible to translate anyway.
Create another branch and cherry-pick the "bold-option" commit, then return to this branch. We will squash commits without it later.

File: src/components/subtitlesettings/subtitleappearancehelper.js [hunk 6]
```diff
@@ -30,13 +30,13 @@ function getTextStyles(settings, preview) {
...
+            list.push({ name: 'text-shadow', value: '0.04em 0.04em #FFF, 0px 0.04em #FFF, 0.04em 0px #FFF, -0.04em -0.04em #000, 0px -0.04em #000, -0.04em 0px #000' });
             break;
         case 'uniform':
-            list.push({ name: 'text-shadow', value: '-1px 0px #000000, 0px 1px #000000, 1px 0px #000000, 0px -1px #000000' });
+            list.push({ name: 'text-shadow', value: '#000 0px 0px 0px, #000 0px 0.025em 0px, #000 0px -0.025em 0px, #000 0px 0.07em 0px, #000 0px -0.07em 0px, #000 0.025em 0px 0px, #000 -0.025em 0px 0px, #000 0.025em 0.025em 0px, #000 -0.025em 0.025em 0px, #000 0.025em -0.025em 0px, #000 -0.025em -0.025em 0px, #000 0.025em 0.07em 0px, #000 -0.025em 0.07em 0px, #000 0.025em -0.07em 0px, #000 -0.025em -0.07em 0px, #000 0.07em 0px 0px, #000 -0.07em 0px 0px, #000 0.07em 0.025em 0px, #000 -0.07em 0.025em 0px, #000 0.07em -0.025em 0px, #000 -0.07em -0.025em 0px, #000 0.07em 0.07em 0px, #000 -0.07em 0.07em 0px, #000 0.07em -0.07em 0px, #000 -0.07em -0.07em 0px, #000 0.09em 0px 0px, #000 -0.09em 0px 0px, #000 0.09em 0.025em 0px, #000 -0.09em 0.025em 0px, #000 0.09em -0.025em 0px, #000 -0.09em -0.025em 0px' });
```
[the remark this answers is not in the document]
Comment (dannymichel) [reply src/components/subtitlesettings/subtitleappearancehelper.js]: https://github.com/jellyfin/jellyfin-web/pull/3607/commits/2635c4b9692c6695d1d9e6ed757c4298e056a412

File: src/components/subtitlesettings/subtitleappearancehelper.js (hunk 5 above)
Comment (dannymichel) [reply src/components/subtitlesettings/subtitleappearancehelper.js]: 56377c7e52b75d7e10bcacdf297d2ee4360ebf31

**Comment by dannymichel** [comment]
Added bold option https://github.com/jellyfin/jellyfin-web/pull/3624

**Comment by dmitrylyzo** [comment]
What they look like now.
subs-em4-wide

**Comment by dannymichel** [comment]
> What they look like now. subs-em4-wide

looks great

**Comment by dannymichel** [comment]
New PR for bold option https://github.com/jellyfin/jellyfin-web/pull/3625

**Comment by dannymichel** [comment]
Everything good, @dmitrylyzo ?

**Comment by dmitrylyzo** [comment]
I think the shadows are still too thick - the contrast hits the eye.
test1

**Comment by dannymichel** [comment]
> I think the shadows are still too thick - the contrast hits the eye. test1

Can you post your version as a commit suggestion as you did above with other lines?

File: src/components/subtitlesettings/subtitleappearancehelper.js [hunk 7]
```diff
@@ -30,13 +30,13 @@ function getTextStyles(settings, preview) {
...
             break;
         case 'depressed':
-            list.push({ name: 'text-shadow', value: '1px 1px white, 0px 1px white, 1px 0px white, -1px -1px black, 0px -1px black, -1px 0px black' });
+            list.push({ name: 'text-shadow', value: '0.04em 0.04em #fff, 0px 0.04em #fff, 0.04em #fff, -0.04em -0.04em #000, 0px -0.04em #000, -0.04em 0px #000' });
             break;
         case 'uniform':
-            list.push({ name: 'text-shadow', value: '-1px 0px #000000, 0px 1px #000000, 1px 0px #000000, 0px -1px #000000' });
+            list.push({ name: 'text-shadow', value: '#000 0px 0.04em, #000 0px -0.04em, #000 0px 0.07em, #000 0px -0.07em, #000 0.04em 0px, #000 -0.04em 0px, #000 0.04em 0.04em, #000 -0.04em 0.04em, #000 0.04em -0.04em, #000 -0.04em -0.04em, #000 0.04em 0.07em, #000 -0.04em 0.07em, #000 0.04em -0.07em, #000 -0.04em -0.07em, #000 0.07em 0px, #000 -0.07em 0px, #000 0.07em 0.04em, #000 -0.07em 0.04em, #000 0.07em -0.04em, #000 -0.07em -0.04em, #000 0.07em 0.07em, #000 -0.07em 0.07em, #000 0.07em -0.07em, #000 -0.07em -0.07em, #000 0.09em 0px, #000 -0.09em 0px, #000 0.09em 0.04em, #000 -0.09em 0.04em, #000 0.09em -0.04em, #000 -0.09em -0.04em' });
```
Comment (dmitrylyzo) [inline src/components/subtitlesettings/subtitleappearancehelper.js:39]: This is the "circle" pattern I was talking about.
```suggestion
            list.push({ name: 'text-shadow', value: '#000 0px 0.03em, #000 0px -0.03em, #000 0px 0.06em, #000 0px -0.06em, #000 0.03em 0px, #000 -0.03em 0px, #000 0.03em 0.03em, #000 -0.03em 0.03em, #000 0.03em -0.03em, #000 -0.03em -0.03em, #000 0.03em 0.06em, #000 -0.03em 0.06em, #000 0.03em -0.06em, #000 -0.03em -0.06em, #000 0.06em 0px, #000 -0.06em 0px, #000 0.06em 0.03em, #000 -0.06em 0.03em, #000 0.06em -0.03em, #000 -0.06em -0.03em' });
```

Comment (dannymichel) [reply src/components/subtitlesettings/subtitleappearancehelper.js]: Ok added the circle pattern https://github.com/jellyfin/jellyfin-web/pull/3607/commits/8058f348b59efe145a177748543f194c77d01986

**Review by dmitrylyzo — APPROVED** [review]
If you need to make it thinner, you should add the old `px`-version which will be used for the small font.

**Review by dmitrylyzo — CHANGES_REQUESTED** [review]
_Just to block merge until rebased._

**Comment by dannymichel** [comment]
> If you need to make it thinner, you should add the old `px`-version which will be used for the small font.

it looks so much better

**Comment by dmitrylyzo** [comment]
> > If you need to make it thinner, you should add the old `px`-version which will be used for the small font.
> 
> it looks so much better

I meant that if you want to go for, say, `0.02em 0.04em`, you can prepend `-1px 0px #000, 0px 1px #000, 1px 0px #000, 0px -1px #000, ` (<- old `px`-version). In this case the `1px`-shadow takes the place of the zeroed `em`-shadow if the font is small.

You have returned useless 0px blur and useless "0,0" shadow.

With the "oval" pattern, the stroke width is uneven, and the flat (horizontal) edge on the `e` is bigger.

**Comment by dmitrylyzo** [comment]
I've played around with the values and I think I like `0.03em 0.05em` more. But it is more a question of rounding `em` to `px`.
subs-em-5

File: src/components/subtitlesettings/subtitleappearancehelper.js (hunk 4 above)
Comment (dannymichel) [reply src/components/subtitlesettings/subtitleappearancehelper.js]: 7e8af63cd7f3f27754f5e3c311a32b69ef93c9c4

File: src/components/subtitlesettings/subtitleappearancehelper.js (hunk 3 above)
Comment (dannymichel) [reply src/components/subtitlesettings/subtitleappearancehelper.js]: 7e8af63cd7f3f27754f5e3c311a32b69ef93c9c4

**Review by dmitrylyzo — APPROVED** [review]

**Pull request merged**